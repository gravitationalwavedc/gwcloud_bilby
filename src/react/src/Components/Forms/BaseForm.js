import React from "react";
import { Form, Button, Grid } from "semantic-ui-react";
import { FormController, FormField } from "./Forms";

class BaseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            valid: false,
            validating: false,
            values: null
        }
    }

    onValid = (valid, values) => {
        this.setState(() => ({
            valid: valid,
            values: values
        }))
    }

    prevStep = () => {
        this.props.prevStep()
    }

    nextStep = () => {
        this.setState(() => ({ validating: true }))
        if (this.state.valid) {
            this.props.updateParentState(this.state.values)
            this.props.nextStep()
        }
    }

    render() {

        return (
            <React.Fragment>
                <Grid.Row>
                    <Grid.Column width={10}>
                        <Form>
                            <Grid textAlign='left'>
                                <FormController
                                    initialValues={this.props.initialData}
                                    validating={this.state.validating}
                                    onValid={this.onValid}
                                >
                                    {
                                        props => {
                                            return (
                                                this.props.setForms(props.values).map(
                                                    (form, index) => (<FormField key={index} {...form} {...props} />)
                                                )
                                            )
                                        }
                                    }
                                </FormController>
                            </Grid>
                        </Form>
                    </Grid.Column>
                </Grid.Row>
                <Grid.Row columns={2}>
                    <Grid.Column floated='left'>
                        {
                            this.props.prevStep && <Button onClick={this.prevStep}>Back</Button>
                        }
                    </Grid.Column>
                    <Grid.Column floated='right'>
                        <Button onClick={this.nextStep}>Save and Continue</Button>
                    </Grid.Column>
                </Grid.Row>
            </React.Fragment>
        )
    }
}

export default BaseForm;