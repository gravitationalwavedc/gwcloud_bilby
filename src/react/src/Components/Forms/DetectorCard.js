import React, { useState } from 'react';
import { Button, Card, Form } from 'react-bootstrap';
import CreatableSelect from 'react-select/creatable';

const activeDetectorStyle = {
    borderLeft: 'none', 
    borderRight: 'none', 
    borderBottom: 'none', 
    borderRadius: 0, 
    boxShadow: '0 4px 8px rgba(0, 0, 0, 0.2)'
};

const deactiveDetectorStyle = {
    borderLeft: 'none', 
    borderRight: 'none', 
    borderBottom: 'none', 
    borderRadius: 0
};

const DetectorCard = ({image, title, formik}) => {
    const [options, setOptions] = useState([
        {label: 'GWOCS', value: 'GWOCS'}, 
        {label: 'GDS-CALIB_STRAIN', value: 'GDS-CALIB_STRAIN'}, 
    ]);

    const identifier = title.toLowerCase();
    const maximumFrequencyId = identifier + 'MaximumFrequency';
    const minimumFrequencyId = identifier + 'MaximumFrequency';
    const channelId = identifier + 'Channel';

    const isActive = formik.values[identifier];

    const toggleActive = () => {
        formik.setFieldValue(identifier, !isActive);
    };
    
    const handleChange = (newValue, actionMeta) => {
        if(actionMeta.action === 'create-option'){
            setOptions([newValue, ...options]);
        }
        formik.setFieldValue(identifier + 'Channel', newValue.value);
    };

    return (
        <Card style={isActive ? activeDetectorStyle: deactiveDetectorStyle}>
            <Card.Img variant="top" src={image} style={{height: '60px', objectFit: 'cover', borderRadius: 0}}/>
            <Card.Header className="h4">
                {title}
                <Button 
                    className="float-right" 
                    variant="outline-primary" 
                    size="sm" 
                    onClick={toggleActive}>
                    {isActive ? 'Deactivate' : 'Activate' }
                </Button>
            </Card.Header>
            <Card.Body>
                <Form.Group controlId={channelId}>
                    <Form.Label>Channel</Form.Label>
                    <CreatableSelect 
                        onChange={handleChange}
                        options={options}
                        value={{value:formik.values[channelId], label: formik.values[channelId]}}
                    />
                </Form.Group>
                <Form.Group controlId={minimumFrequencyId}>
                    <Form.Label>Minimum Frequency</Form.Label>
                    <Form.Control 
                        name={minimumFrequencyId} 
                        {...formik.getFieldProps(minimumFrequencyId)}/>
                </Form.Group>
                <Form.Group controlId={maximumFrequencyId}>
                    <Form.Label>Maximum Frequency</Form.Label>
                    <Form.Control 
                        name={maximumFrequencyId} 
                        {...formik.getFieldProps(maximumFrequencyId)}/>
                </Form.Group>
            </Card.Body>
        </Card>
    );
};


export default DetectorCard;
