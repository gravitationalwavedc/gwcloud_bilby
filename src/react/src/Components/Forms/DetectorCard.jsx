import { useState, useContext } from 'react';
import { Button, Card, Form } from 'react-bootstrap';
import CreatableSelect from 'react-select/creatable';
import Input from './Atoms/Input';
import { isLigoUser } from '../../Utils/UserUtils';
import { UserContext } from '../../sessionUser';

const DetectorCard = ({ image, title, formik, channelOptions }) => {
    const [options, setOptions] = useState(channelOptions);
    const user = useContext(UserContext);

    const identifier = title.toLowerCase();
    const maximumFrequencyId = identifier + 'MaximumFrequency';
    const minimumFrequencyId = identifier + 'MinimumFrequency';
    const channelId = identifier + 'Channel';

    const isActive = formik.values[identifier];

    const toggleActive = () => {
        formik.setFieldValue(identifier, !isActive);
    };

    const channelSelectValue = isLigoUser(user)
        ? {
              value: formik.values[channelId],
              label: formik.values[channelId],
          }
        : {
              value: 'GWOSC',
              label: 'GWOSC',
          };

    const handleChange = (newValue, actionMeta) => {
        if (actionMeta.action === 'create-option') {
            setOptions([newValue, ...options]);
        }
        formik.setFieldValue(identifier + 'Channel', newValue.value);
    };

    return (
        <Card className={isActive ? 'gw-detector-card active' : 'gw-detector-card'}>
            <Card.Img variant="top" src={image} />
            <Card.Header className="h4">
                {title}
                <Button
                    data-testid={identifier + 'Active'}
                    className="float-right"
                    variant="outline-primary"
                    size="sm"
                    onClick={toggleActive}
                >
                    {isActive ? 'Deactivate' : 'Activate'}
                </Button>
            </Card.Header>
            <Card.Body>
                <Form.Group controlId={channelId}>
                    <Form.Label>Channel</Form.Label>
                    <CreatableSelect
                        className="gw-select"
                        classNamePrefix="gw-select"
                        isDisabled={!isActive || !isLigoUser(user)}
                        onChange={handleChange}
                        options={options}
                        value={channelSelectValue}
                    />
                    <Form.Text id={channelId + 'Help'}>Start typing for a custom channel.</Form.Text>
                </Form.Group>
                <Input
                    formik={formik}
                    title="Minimum frequency"
                    name={minimumFrequencyId}
                    type="number"
                    disabled={!isActive}
                />
                <Input
                    formik={formik}
                    title="Maximum frequency"
                    name={maximumFrequencyId}
                    type="number"
                    disabled={!isActive}
                />
            </Card.Body>
        </Card>
    );
};

export default DetectorCard;
