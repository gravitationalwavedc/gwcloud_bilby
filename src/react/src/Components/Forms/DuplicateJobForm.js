import React from 'react';
import {graphql, createFragmentContainer} from 'react-relay';
import NewJob from '../../Pages/NewJob';
import parseJobParams from '../../Utils/ParseJobParams';

const DuplicateJobForm = (props) => {
    const params = props.data.bilbyJob.params;

    const initialValues = parseJobParams(params);

    initialValues['name'] = `Copy-of-${initialValues.name}`;
    initialValues['description'] = `A duplicate job of ${initialValues.name}. Original description: ${initialValues.description}`;

    return <NewJob initialValues={initialValues} {...props}/>;
};

export default createFragmentContainer(DuplicateJobForm,
    {
        data: graphql`
        fragment DuplicateJobForm_data on Query @argumentDefinitions(
            jobId: {type: "ID!"}
        ){
            bilbyJob (id: $jobId){
                params {
                    details {
                        name
                        description
                        private
                    }
                    data {
                        dataChoice
                        triggerTime
                        channels {
                            hanfordChannel
                            livingstonChannel
                            virgoChannel
                        }
                    }
                    detector {
                        hanford
                        hanfordMinimumFrequency
                        hanfordMaximumFrequency
                        livingston
                        livingstonMinimumFrequency
                        livingstonMaximumFrequency
                        virgo
                        virgoMinimumFrequency
                        virgoMaximumFrequency
                        duration
                        samplingFrequency
                    }
                    prior {
                        priorDefault
                    }
                    sampler {
                        nlive
                        nact
                        maxmcmc
                        walks
                        dlogz
                        cpus
                        samplerChoice
                    }
                    waveform {
                        model
                    }
                }
           }
           
           ...DataForm_data
        }`
    }
);
