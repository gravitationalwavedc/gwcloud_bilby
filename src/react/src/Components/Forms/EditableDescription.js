import React from 'react';
import EditableText from './EditableText';
import { graphql } from 'react-relay';
import { commitMutation } from 'relay-runtime';
import {harnessApi} from '../../index';

const mutation = graphql`
    mutation EditableDescriptionMutation($input: UpdateBilbyJobMutationInput!) {
        updateBilbyJob(input: $input) {
            result
            jobId
        }
    }
`;

const handleSaveDescription = (value, jobId) => {
    const variables = {
        input: {
            jobId: jobId,
            description: value 
        }
    };

    commitMutation(harnessApi.getEnvironment('bilby'), {
        mutation: mutation,
        variables: variables,
        onCompleted: (response, errors) => {
            if (!errors) {
                console.log('saved!');
            }
        }
    });
};

const EditableDescription = ({modifiable, value, jobId}) => <React.Fragment>
    {modifiable ? 
        <EditableText 
            name="description" 
            type="textarea"
            value={value} 
            onSave={(value) => handleSaveDescription(value, jobId)} 
            viewProps={{className: 'p'}}
        /> 
        : <p>{value}</p>}
</React.Fragment>;

export default EditableDescription;
