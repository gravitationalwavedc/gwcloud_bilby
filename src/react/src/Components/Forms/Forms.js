import React from "react";
import _ from "lodash";
import {Form, Divider, Grid, Label} from "semantic-ui-react";
import {assembleErrorString} from "../../Utils/errors";


// Might need to be able to store visibility functions to calculate in handleChange

const FormContext = React.createContext()

class FormController extends React.Component {
    constructor(props) {
        super(props);

        const initialErrors = {}
        Object.keys(this.props.initialValues).map((key) => {
            initialErrors[key] = []
        })

        const initialVisible = {}
        Object.keys(this.props.initialValues).map((key) => {
            initialVisible[key] = true
        })

        this.state = {
            values: this.props.initialValues || {},
            errors: initialErrors,
            visible: initialVisible,
            touched: {},
            isValidating : false,
            isValid: false
        }
    }

    handleChange = (values, errors) => {
        this.setState(prevState => ({
            values: {
                ...prevState.values,
                ...values
            },
            errors: {
                ...prevState.errors,
                ...errors
            }
        }), () => this.setValid())
    }

    handleVisibility = (name, visible) => {
        this.setState(prevState => ({
            visible: {
                ...prevState.visible,
                [name]: visible
            }
        }))
    }

    setValid = () => {
        const notEmpty = (arr) => {return Boolean(arr && arr.length)}
        const visibleErrors = _.pick(this.state.errors, Object.keys(_.pickBy(this.state.visible, val => val)))
        
        this.setState({
            isValid: !Object.values(visibleErrors).some(notEmpty)
        })
    }

    setValidating = () => {
        this.setState({
            isValidating: true
        })
    }
        
    render() {
        return <FormContext.Provider value={{
            ...this.state,
            handleChange: this.handleChange,
            handleVisibility: this.handleVisibility,
            setValid: this.setValid,
            setInitial: this.setInitial
        }}>
            {
                this.props.children({
                    ...this.state
                })
            }
        </FormContext.Provider>
    }
}


class FormField extends React.Component {
    
    static defaultProps = {
        visible: true,
        required: true,
        valFunc: () => {},
        errFunc: () => []
    }

    componentDidMount() {
        const { name, visible } = this.props
        const { values, handleVisibility, handleChange } = this.context

        const [updatedValues, errors] = this._handleUpdatingValuesAndSettingErrors(name, values[name])

        handleVisibility(name, visible)
        handleChange(updatedValues, errors)
    }

    componentDidUpdate(prevProps) {
        const { name, visible } = this.props
        
        if (visible !== prevProps.visible) {
            this.context.handleVisibility(name, visible)
        }
    }

    handleChange = (e, data) => {
        const value = data.type === 'checkbox' ? data.checked : data.value
        const { name } = data
        
        const [values, errors] = this._handleUpdatingValuesAndSettingErrors(name, value)

        this.context.handleChange(values, errors)
    }
    
    _handleUpdatingValuesAndSettingErrors(name, value) {
        const {errFunc, linkedErrors, valFunc} = this.props

        const values = {
            [name]: value,
            ...valFunc(value)
        }

        var errors = {
            [name]: errFunc(values[name]),
            ...(_.mapValues(linkedErrors, (linkedErrFunc, linkedErrName) => linkedErrFunc(value)(this.context.values[linkedErrName])))
        }


        return [values, errors]
    }

    render() {
        const { label, name, form, required, visible } = this.props
        const value = this.context.values[name]
        const errors = this.context.errors[name]
        const { isValidating } = this.context
        
        let showError = false
        if (errors && errors.length && isValidating) {
            showError = true
        }

        const childProps = {onChange: this.handleChange, name: name, error: showError}
        if (form.props.control && form.props.control.name === 'Checkbox') {
            childProps['checked'] = value
        } else {
            childProps['value'] = value
        }

        return (
            visible ?
                <Grid.Row columns={3}>
                    <Grid.Column verticalAlign='middle' width={4}>
                        <Form.Field required={required} label={label}/>
                    </Grid.Column>
                    <Grid.Column verticalAlign='middle' width={8}>
                        {React.cloneElement(form, childProps)}
                    </Grid.Column>
                    <Grid.Column verticalAlign='middle' width={4}>
                        {showError ? <Label basic pointing='left' color='red' content={assembleErrorString(errors)}/> : null}
                    </Grid.Column>
                </Grid.Row>
            :
                null
        )

    }
}
FormField.contextType = FormContext

export {FormController, FormField, FormContext}