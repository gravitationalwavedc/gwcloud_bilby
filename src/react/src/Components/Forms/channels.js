const hanford = [
  { label: "GWOSC", value: "GWOSC" },
  { label: "GDS-CALIB_STRAIN", value: "GDS-CALIB_STRAIN" },
];

const livingston = [
  { label: "GWOSC", value: "GWOSC" },
  { label: "GDS-CALIB_STRAIN", value: "GDS-CALIB_STRAIN" },
];

const virgo = [
  { label: "GWOSC", value: "GWOSC" },
  { label: "Hrec_hoft_16384Hz", value: "Hrec_hoft_16384Hz" },
];

export { hanford, livingston, virgo };
