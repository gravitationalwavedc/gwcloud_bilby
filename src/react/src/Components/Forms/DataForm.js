import React from "react";
import {FormController, FormField} from "./Forms";
import {Form, Grid, Button, Select} from "semantic-ui-react";
import {checkForErrors, isNumber, notEmpty, noneFalse} from "../../Utils/errors";
import _ from "lodash";

import {graphql, createFragmentContainer} from "react-relay";
import * as Enumerable from "linq";

class DataForm extends React.Component {
    constructor(props) {
        super(props);
        this.formController = React.createRef();

        const channelOptions = [
            'GWOSC',
            'GDS-CALIB_STRAIN',
            'Hrec_hoft_16384Hz',
        ]
        
        this.state = {
            channelOptions: _.uniq(channelOptions)
        }
    }

    handleAddition = (e, { value }) => {
        this.setState((prevState) => ({
          channelOptions: [value, ...prevState.channelOptions],
        }))
    }

    prevStep = () => {
        this.props.prevStep()
    }

    nextStep = () => {
        this.formController.current.setValidating()
        if (this.formController.current.state.isValid) {
            this.props.updateParentState(this.formController.current.state.values)
            this.props.nextStep()
        }
    }

    setForms(values) {
        const forms = [
            {
                label: 'Type of Data',
                name: 'dataChoice',
                form: <Form.Select placeholder="Select Data Type" options={
                    [
                        {key: 'open', text: 'Open', value: 'open'},
                        {key: 'simulated', text: 'Simulated', value: 'simulated'}
                    ]
                }/>,
                valFunc: (val) => val === 'open' ? {hanfordChannel: 'GWOSC', livingstonChannel: 'GWOSC', virgoChannel: 'GWOSC'} : {},
            },

            {
                label: 'Detectors',
                name: 'hanford',
                form: <Form.Checkbox label="Hanford"/>,
                errFunc: checkForErrors(noneFalse([values.livingston, values.virgo])),
                linkedErrors: {
                    livingston: (val) => checkForErrors(noneFalse([val, values.virgo])),
                    virgo: (val) => checkForErrors(noneFalse([val, values.livingston]))
                }
            },

            {
                label: null,
                name: 'livingston',
                form: <Form.Checkbox label="Livingston"/>,
                errFunc: checkForErrors(noneFalse([values.hanford, values.virgo])),
                linkedErrors: {
                    hanford: (val) => checkForErrors(noneFalse([val, values.virgo])),
                    virgo: (val) => checkForErrors(noneFalse([val, values.livingston]))
                }
            },

            {
                label: null,
                name: 'virgo',
                form: <Form.Checkbox label="Virgo"/>,
                errFunc: checkForErrors(noneFalse([values.hanford, values.livingston])),
                linkedErrors: {
                    hanford: (val) => checkForErrors(noneFalse([values.livingston, val])),
                    livingston: (val) => checkForErrors(noneFalse([values.hanford, val]))
                }
            },

            {
                label: 'Signal Duration (s)',
                name: 'signalDuration',
                form: <Form.Select placeholder="Select Signal Duration" options={
                    [
                        {key: '4', text: '4', value: '4'},
                        {key: '8', text: '8', value: '8'},
                        {key: '16', text: '16', value: '16'},
                        {key: '32', text: '32', value: '32'},
                        {key: '64', text: '64', value: '64'},
                        {key: '128', text: '128', value: '128'}
                    ]
                }/>
            },

            {
                label: 'Sampling Frequency (Hz)',
                name: 'samplingFrequency',
                form: <Form.Select placeholder="Select Sampling Frequency" options={
                    [
                        {key: '512', text: '512', value: '512'},
                        {key: '1024', text: '1024', value: '1024'},
                        {key: '2048', text: '2048', value: '2048'},
                        {key: '4096', text: '4096', value: '4096'},
                        {key: '8192', text: '8192', value: '8192'},
                        {key: '16384', text: '16384', value: '16384'},
                    ]
                }/>
            },

            {
                label: 'Trigger Time',
                name: 'triggerTime',
                form: <Form.Input placeholder='2.1'/>,
                errFunc: checkForErrors(isNumber, notEmpty)
            },
        ]
        

        Enumerable.from(['hanford', 'livingston', 'virgo']).forEach((name) => {
            forms.push(
                {
                    label: name[0].toUpperCase() + name.slice(1) + ': Minimum Frequency (Hz)',
                    name: name + 'MinimumFrequency',
                    form: <Form.Input placeholder=''/>,
                    errFunc: checkForErrors(isNumber, notEmpty),
                    visible: values[name]
                },

                {
                    label: name[0].toUpperCase() + name.slice(1) + ': Maximum Frequency (Hz)',
                    name: name + 'MaximumFrequency',
                    form: <Form.Input placeholder=''/>,
                    errFunc: checkForErrors(isNumber, notEmpty),
                    visible: values[name]
                },
                
                {
                    label: name[0].toUpperCase() + name.slice(1) + ': Channel',
                    name: name + 'Channel',
                    form: <Form.Field 
                        control={Select} 
                        search 
                        allowAdditions 
                        onAddItem={this.handleAddition} 
                        placeholder="Select Channel"
                        options={
                                // Create options from list of values, excluding some of the values
                            _.without(this.state.channelOptions, name==='virgo' ? 'GDS-CALIB_STRAIN' : 'Hrec_hoft_16384Hz').map(value => ({key: value, text: value, value: value}))
                        }
                        disabled={values.dataChoice==='open'}
                    />,
                    errFunc: checkForErrors(notEmpty),
                    visible: values[name]

                }
            )
        })

        return forms
    }

    render() {
        var initialData = {
            dataChoice: 'open',
            hanford: true, // This is because of an annoying bug with errors.. I may need to, once again, refactor the forms.
            livingston: false,
            virgo: false,
            signalDuration: '4',
            samplingFrequency: '16384',
            triggerTime: '',
            hanfordMinimumFrequency: '20',
            hanfordMaximumFrequency: '1024',
            hanfordChannel: 'GWOSC',
            livingstonMinimumFrequency: '20',
            livingstonMaximumFrequency: '1024',
            livingstonChannel: 'GWOSC',
            virgoMinimumFrequency: '20',
            virgoMaximumFrequency: '1024',
            virgoChannel: 'GWOSC'
        }

        initialData = (this.props.data !== null) ? this.props.data : initialData
        initialData = (this.props.state !== null) ? this.props.state : initialData

        return (
            <React.Fragment>
                <FormController
                    initialValues={initialData}
                    ref={this.formController}
                >
                    {
                        props => {
                            return (
                                <Grid.Row>
                                    <Grid.Column width={10}>
                                        <Form>
                                            <Grid textAlign='left'>
                                                {
                                                    this.setForms(props.values).map(
                                                        (form, index) => (<FormField key={index} {...form}/>)
                                                    )
                                                }
                                            </Grid>
                                        </Form>
                                    </Grid.Column>
                                </Grid.Row>
                            )
                        }
                    }
                </FormController>
                <Grid.Row columns={2}>
                    <Grid.Column floated='left'>
                        <Button onClick={this.prevStep}>Back</Button>
                    </Grid.Column>
                    <Grid.Column floated='right'>
                        <Button onClick={this.nextStep}>Save and Continue</Button>
                    </Grid.Column>
                </Grid.Row>
            </React.Fragment>
        )
    }
}

export default createFragmentContainer(DataForm, {
    data: graphql`
        fragment DataForm_data on DataType {
            dataChoice
            hanford
            livingston
            virgo
            signalDuration
            samplingFrequency
            triggerTime
            hanfordMinimumFrequency
            hanfordMaximumFrequency
            hanfordChannel
            livingstonMinimumFrequency
            livingstonMaximumFrequency
            livingstonChannel
            virgoMinimumFrequency
            virgoMaximumFrequency
            virgoChannel
        }
    `
});