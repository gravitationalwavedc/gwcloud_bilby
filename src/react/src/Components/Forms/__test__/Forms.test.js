// Set the harnessApi
import React from "react";
import {expect} from "@jest/globals";
import { FormController, FormField, FormContext } from "../Forms";
import { Form, Grid } from "semantic-ui-react";
import { mount } from "enzyme";
import { checkForErrors, assembleErrorString, shorterThan } from "../../../Utils/errors";


describe('Form Controller with Form Field', () => {
    let tree = (
        <FormController
            initialValues={{test: 'Test'}}
        >
            {
                props => {
                    return (
                        <Form>
                            <Grid textAlign='left'>
                                <FormField name='test' label='Test' form={<Form.Input />} errFunc={checkForErrors(shorterThan(5))} visible={props.values.test !== 'Invisible'}/>
                            </Grid>
                        </Form>
                    )
                }
            }
        </FormController>
    )

    let mountedTree = mount(tree)
    let inputField = mountedTree.find('input')
    
    it('renders correctly', () => {
        expect(mountedTree).toMatchSnapshot();
    })

    it('places initial value in the form field', () => {
        expect(inputField.prop('value')).toBe('Test')
    })

    it('modifies FormController state on change', () => {
        inputField.instance().value = 'Testing'
        inputField.simulate('change')
        
        expect(mountedTree.state().values.test).toBe('Testing')
        expect(mountedTree.state().errors.test).toHaveLength(1)
    })

    it('displays error message when set to validate', () => {
        mountedTree.instance().setValidating()
        mountedTree.update()

        expect(mountedTree.find('Label').get(0).props.content).toBe(assembleErrorString(checkForErrors(shorterThan(5))('Testing')))
        expect(mountedTree).toMatchSnapshot()
    })
    
    it('vanishes when visible is false', () => {
        inputField.instance().value = 'Invisible'
        inputField.simulate('change')
        
        expect(mountedTree.state().visible.test).toBe(false)
        expect(mountedTree).toMatchSnapshot()
    })

})