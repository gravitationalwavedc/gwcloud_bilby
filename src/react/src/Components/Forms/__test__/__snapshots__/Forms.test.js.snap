// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Controller with Form Field displays error message when set to validate 1`] = `
<FormController
  initialValues={
    Object {
      "test": "Test",
    }
  }
>
  <Form
    as="form"
  >
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <Grid
        textAlign="left"
      >
        <div
          className="ui left aligned grid"
        >
          <FormField
            errFunc={[Function]}
            form={
              <FormInput
                as={[Function]}
                control={[Function]}
              />
            }
            label="Test"
            name="test"
            required={true}
            valFunc={[Function]}
            visible={true}
          >
            <GridRow
              columns={3}
            >
              <div
                className="three column row"
              >
                <GridColumn
                  verticalAlign="middle"
                  width={4}
                >
                  <div
                    className="middle aligned four wide column"
                  >
                    <FormField
                      label="Test"
                      required={true}
                    >
                      <div
                        className="required field"
                      >
                        <label>
                          Test
                        </label>
                      </div>
                    </FormField>
                  </div>
                </GridColumn>
                <GridColumn
                  verticalAlign="middle"
                  width={8}
                >
                  <div
                    className="middle aligned eight wide column"
                  >
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      error={true}
                      name="test"
                      onChange={[Function]}
                      value="Testing"
                    >
                      <FormField
                        control={[Function]}
                        error={true}
                        name="test"
                        onChange={[Function]}
                        value="Testing"
                      >
                        <div
                          className="error field"
                        >
                          <Input
                            aria-describedby={null}
                            aria-invalid={true}
                            name="test"
                            onChange={[Function]}
                            type="text"
                            value="Testing"
                          >
                            <div
                              className="ui input"
                            >
                              <input
                                aria-describedby={null}
                                aria-invalid={true}
                                name="test"
                                onChange={[Function]}
                                type="text"
                                value="Testing"
                              />
                            </div>
                          </Input>
                        </div>
                      </FormField>
                    </FormInput>
                  </div>
                </GridColumn>
                <GridColumn
                  verticalAlign="middle"
                  width={4}
                >
                  <div
                    className="middle aligned four wide column"
                  >
                    <Label
                      basic={true}
                      color="red"
                      content="Must be shorter than 5 characters"
                      pointing="left"
                    >
                      <div
                        className="ui red left pointing basic label"
                        onClick={[Function]}
                      >
                        Must be shorter than 5 characters
                      </div>
                    </Label>
                  </div>
                </GridColumn>
              </div>
            </GridRow>
          </FormField>
        </div>
      </Grid>
    </form>
  </Form>
</FormController>
`;

exports[`Form Controller with Form Field renders correctly 1`] = `
<FormController
  initialValues={
    Object {
      "test": "Test",
    }
  }
>
  <Form
    as="form"
  >
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <Grid
        textAlign="left"
      >
        <div
          className="ui left aligned grid"
        >
          <FormField
            errFunc={[Function]}
            form={
              <FormInput
                as={[Function]}
                control={[Function]}
              />
            }
            label="Test"
            name="test"
            required={true}
            valFunc={[Function]}
            visible={true}
          >
            <GridRow
              columns={3}
            >
              <div
                className="three column row"
              >
                <GridColumn
                  verticalAlign="middle"
                  width={4}
                >
                  <div
                    className="middle aligned four wide column"
                  >
                    <FormField
                      label="Test"
                      required={true}
                    >
                      <div
                        className="required field"
                      >
                        <label>
                          Test
                        </label>
                      </div>
                    </FormField>
                  </div>
                </GridColumn>
                <GridColumn
                  verticalAlign="middle"
                  width={8}
                >
                  <div
                    className="middle aligned eight wide column"
                  >
                    <FormInput
                      as={[Function]}
                      control={[Function]}
                      error={false}
                      name="test"
                      onChange={[Function]}
                      value="Test"
                    >
                      <FormField
                        control={[Function]}
                        error={false}
                        name="test"
                        onChange={[Function]}
                        value="Test"
                      >
                        <div
                          className="field"
                        >
                          <Input
                            aria-describedby={null}
                            aria-invalid={true}
                            name="test"
                            onChange={[Function]}
                            type="text"
                            value="Test"
                          >
                            <div
                              className="ui input"
                            >
                              <input
                                aria-describedby={null}
                                aria-invalid={true}
                                name="test"
                                onChange={[Function]}
                                type="text"
                                value="Test"
                              />
                            </div>
                          </Input>
                        </div>
                      </FormField>
                    </FormInput>
                  </div>
                </GridColumn>
                <GridColumn
                  verticalAlign="middle"
                  width={4}
                >
                  <div
                    className="middle aligned four wide column"
                  />
                </GridColumn>
              </div>
            </GridRow>
          </FormField>
        </div>
      </Grid>
    </form>
  </Form>
</FormController>
`;

exports[`Form Controller with Form Field vanishes when visible is false 1`] = `
<FormController
  initialValues={
    Object {
      "test": "Test",
    }
  }
>
  <Form
    as="form"
  >
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <Grid
        textAlign="left"
      >
        <div
          className="ui left aligned grid"
        >
          <FormField
            errFunc={[Function]}
            form={
              <FormInput
                as={[Function]}
                control={[Function]}
              />
            }
            label="Test"
            name="test"
            required={true}
            valFunc={[Function]}
            visible={false}
          />
        </div>
      </Grid>
    </form>
  </Form>
</FormController>
`;
